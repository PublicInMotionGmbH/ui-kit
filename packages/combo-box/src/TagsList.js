import React from 'react'
import PropTypes from 'prop-types'

import { buildClassName } from '@talixo/shared'

import Tag from './Tag'

/**
 * Component which represents list of value tags for multi-select elements
 *
 * @param {object} props
 * @param {function} props.buildItemId
 * @param {array} props.selectedItems
 *
 * @returns {React.Element}
 */
function TagsList (props) {
  const { buildItemId, selectedItems, style } = props

  // Build class name for wrapper
  const clsName = buildClassName([ 'combo-box', 'tags' ])

  // Build elements for all selected items inside
  const elements = selectedItems.map((item, index) => (
    <Tag
      key={buildItemId(item, index)}
      {...props}
      item={item}
      index={index}
    />
  ))

  return (
    <div style={style} className={clsName}>
      {elements}
    </div>
  )
}

TagsList.propTypes = {
  /** Additional styles for wrapper */
  style: PropTypes.object,

  /** Function to build item ID - used for 'key' properties */
  buildItemId: PropTypes.func.isRequired,

  /** List of currently selected items */
  selectedItems: PropTypes.array,

  /** Function to build 'remove' button props with handlers */
  getRemoveButtonProps: PropTypes.func.isRequired,

  /** Function to render item value */
  renderValue: PropTypes.func.isRequired
}

TagsList.defaultProps = {
  selectedItems: []
}

export default TagsList
