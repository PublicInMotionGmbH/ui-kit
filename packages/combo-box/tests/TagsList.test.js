import React from 'react'
import { shallow } from 'enzyme'

import TagsList from '../src/TagsList'

describe('<TagsList />', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <TagsList
        buildItemId={item => item}
        getRemoveButtonProps={() => {}}
        renderValue={item => item}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render passed items', () => {
    const render = item => item
    const getRemoveButtonProps = () => ({})

    const wrapper = shallow(
      <TagsList
        selectedItems={[ 'abc', 'def', 'ghi' ]}
        buildItemId={item => item}
        getRemoveButtonProps={getRemoveButtonProps}
        renderValue={render}
      />
    )

    expect(wrapper.children().length).toBe(3)
    expect(wrapper.children().at(0).prop('item')).toBe('abc')
    expect(wrapper.children().at(1).prop('item')).toBe('def')
    expect(wrapper.children().at(2).prop('item')).toBe('ghi')
  })

  it('should pass correct props to items', () => {
    const render = item => item
    const getRemoveButtonProps = () => ({})

    const wrapper = shallow(
      <TagsList
        selectedItems={[ 'abc', 'def', 'ghi' ]}
        buildItemId={item => item}
        getRemoveButtonProps={getRemoveButtonProps}
        renderValue={render}
      />
    )

    expect(wrapper.children().at(0).prop('renderValue')).toBe(render)
    expect(wrapper.children().at(0).prop('getRemoveButtonProps')).toBe(getRemoveButtonProps)
  })

  it('should build keys for items', () => {
    const render = item => item
    const getRemoveButtonProps = () => ({})

    const wrapper = shallow(
      <TagsList
        selectedItems={[ 'abc', 'def', 'ghi' ]}
        buildItemId={(item, index) => index}
        getRemoveButtonProps={getRemoveButtonProps}
        renderValue={render}
      />
    )

    expect(wrapper.children().at(0).key()).toBe('0')
    expect(wrapper.children().at(1).key()).toBe('1')
    expect(wrapper.children().at(2).key()).toBe('2')

    wrapper.setProps({ buildItemId: item => item })

    expect(wrapper.children().at(0).key()).toBe('abc')
    expect(wrapper.children().at(1).key()).toBe('def')
    expect(wrapper.children().at(2).key()).toBe('ghi')
  })
})
