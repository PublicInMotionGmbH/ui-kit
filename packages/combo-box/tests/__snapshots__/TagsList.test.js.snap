// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TagsList /> should render correctly 1`] = `
<div
  className="talixo-combo-box__tags"
/>
`;
